### PR TITLE
Add a Windows build script

### PR DESCRIPTION
**Pull request by chambbj**

### CODE REVIEW

**Comment by chambbj** [comment]
@conda-forge-admin, please rerender

**Comment by chambbj** [comment]
Wow, how did I not notice that there was already a PR for Windows support before starting this! Looks like it's been a few years though, so maybe a fresh start will do the trick. ;)

**Comment by chambbj** [comment]
@conda-forge/flann I believe this should take care of the Windows build for `flann` and can supersede #1. Happy to hear your thoughts.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -8,29 +8,44 @@ source:
...
-    number: 0
-    skip: true    # [win]
+    number: 1
+    features:
+        - vc9   # [win and py27]
+        - vc14  # [win and py>=35]
 
 requirements:
     build:
         - gcc     # [unix]
```
Comment (jakirkham) [inline recipe/meta.yaml]: Please add the following...

```yaml
        - {{ compiler('cxx') }}     # [win]
```

Once added all `vc` features and `vc` dependencies can be dropped.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -8,29 +8,44 @@ source:
...
+        - python  # [win]
```
Comment (jakirkham) [inline recipe/meta.yaml]: Also can drop `python`.

**Comment by jakirkham** [comment]
Thanks for the submission @chambbj. Looks pretty good.

The compiler situation on Windows is changing significantly because of `conda-build` 3. This will ultimately affect all platforms, but not right way. Have added some notes above on how to adjust this.

Docs are being discussed in PR ( https://github.com/conda-forge/conda-forge.github.io/pull/576 ).

**Comment by chambbj** [comment]
Thanks for the link and suggestions @jakirkham. I knew there had been some update to all of this, but couldn't remember where to find them. It does make for a much cleaner meta.yaml!

File: recipe/meta.yaml [hunk 3]
```diff
@@ -8,14 +8,16 @@ source:
     fn: flann-{{ version }}.tar.gz
     url: https://github.com/mariusmuja/flann/archive/{{ version }}.tar.gz
     sha256: b23b5f4e71139faa3bcb39e6bbcc76967fbaf308c4ee9d4f5bfbeceaa76cc5d3
+    patches:
+        - cmake.patch  # [linux]
```
Comment (djsutherland) [inline recipe/meta.yaml]: Can we give this a more informative name and/or a comment about why it’s needed?

**Comment by djsutherland** [comment]
Thanks @chambbj, this looks good. :)

So we're not actually testing anything here, just that it compiled, which is a little bit worrisome. This is partly because flann's tests are hard to run (https://github.com/mariusmuja/flann/issues/113). Added some `conda inspect` stuff to at least make sure that the binaries are reasonably valid, but that doesn't actually work on Windows anyway....

File: recipe/cmake.patch [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+-
+-    if (BUILD_CUDA_LIB)
+-	    cuda_add_library(flann_cuda SHARED "")
+-        set_target_properties(flann_cuda PROPERTIES LINKER_LANGUAGE CXX)
+-        target_link_libraries(flann_cuda -Wl,-whole-archive flann_cuda_s -Wl,-no-whole-archive)
+-        set_property(TARGET flann_cpp_s PROPERTY COMPILE_DEFINITIONS FLANN_USE_CUDA)
+-    # 	target_link_libraries(flann_cuda cudpp_x86_64)
+-    endif()
+-else()
++#if(CMAKE_SYSTEM_NAME STREQUAL "Linux" AND CMAKE_COMPILER_IS_GNUCC)
```
Comment (jakirkham) [inline recipe/cmake.patch]: What's the value of commenting this on Windows? Based on the condition here it seems like this whole section would be skipped anyways and the `else` below run even on Windows.

File: recipe/cmake.patch [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+-    if(CMAKE_SYSTEM_NAME STREQUAL "Linux" AND CMAKE_COMPILER_IS_GNUCC)
+-        add_library(flann SHARED "")
+-        set_target_properties(flann PROPERTIES LINKER_LANGUAGE CXX)
+-        target_link_libraries(flann -Wl,-whole-archive flann_s -Wl,-no-whole-archive)
+-    else()
++    #if(CMAKE_SYSTEM_NAME STREQUAL "Linux" AND CMAKE_COMPILER_IS_GNUCC)
++    #    add_library(flann SHARED "")
++    #    set_target_properties(flann PROPERTIES LINKER_LANGUAGE CXX)
++    #    target_link_libraries(flann -Wl,-whole-archive flann_s -Wl,-no-whole-archive)
++    #else()
```
Comment (jakirkham) [inline recipe/cmake.patch]: Same question as above.

File: recipe/meta.yaml (hunk 3 above)
Comment (jakirkham) [reply recipe/meta.yaml]: After looking at the patch more closely, am not clear that it is actually needed. :/

File: recipe/cmake.patch (hunk 4 above)
Comment (djsutherland) [reply recipe/cmake.patch]: Note that this patch is applied on Linux, not on Windows.

I don't entirely know what the point is and it'd be good to clarify. :)

File: recipe/meta.yaml (hunk 3 above)
Comment (chambbj) [reply recipe/meta.yaml]: I will add a comment and rename. This is a known issue though as reported at https://github.com/mariusmuja/flann/issues/369, and as evidenced at https://circleci.com/gh/conda-forge/flann-feedstock/35 (this PR, prior to adding the patch).

File: recipe/cmake.patch (hunk 4 above)
Comment (chambbj) [reply recipe/cmake.patch]: As noted elsewhere, a patch is required to due to the way CMake 3.11 treats `add_library` commands with empty strings. That said, perhaps adding empty source files, as suggested at https://github.com/mariusmuja/flann/issues/369 may be a better workaround. Regardless, I'll document this issue more clearly in the meta.yaml.

**Comment by chambbj** [comment]
@jakirkham @dougalsutherland Hopefully the latest commits make the required patch a little more clear. Let me know if there is anything else I need to address, or if you'd like me to squash things.

File: recipe/cmake.patch (hunk 4 above)
Comment (jakirkham) [reply recipe/cmake.patch]: I see. So somewhere later on they are adding things to include in the construction of these libraries, but they need to initialize them first.

**Comment by jakirkham** [comment]
Thanks @chambbj. That makes more sense.

Given we are using a newer CMake, we can probably get away with just calling `add_library` with no arguments for the source (instead of an empty string) and just dropping the `empty.cpp` file. This sounds like CMake's intended usage.

File: recipe/cmake.patch (hunk 4 above)
Comment (jakirkham) [reply recipe/cmake.patch]: Sorry must have overlooked the selector.

Comment (chambbj) [reply recipe/cmake.patch]: Actually, I think they are creating the static library first and then trying to create the shared library directly from the static, but when CMake added this ability to add source files to the initialized (but empty) target, it messed up that logic.

**Comment by chambbj** [comment]
@jakirkham I'd have to rerun to give you the exact error, but this actually doesn't work in this case. See my other note about how they are actually creating their shared library from the static one (this is to save compilation time).

File: recipe/meta.yaml [hunk 6]
```diff
@@ -24,13 +28,21 @@ requirements:
...
+        - test -f $PREFIX/lib/libflann.dylib                  # [osx]
+        - test -f $PREFIX/lib/libflann.so                     # [linux]
+        - test -f $PREFIX/lib/libflann_cpp_s.a                # [unix]
+        - test -f $PREFIX/lib/libflann_cpp.dylib              # [osx]
+        - test -f $PREFIX/lib/libflann_cpp.so                 # [linux]
+        - if not exist %LIBRARY_LIB%\\flann_cpp_s.lib exit 1  # [win]
+        - if not exist %LIBRARY_LIB%\\flann_s.lib exit 1      # [win]
+        - if not exist %LIBRARY_LIB%\\flann.lib exit 1        # [win]
+        - if not exist %LIBRARY_BIN%\\flann.dll exit 1        # [win]
+        - if not exist %LIBRARY_INC%\\flann exit 1            # [win]
```
Comment (jakirkham) [inline recipe/meta.yaml]: We could tidy these a bit to keep the easier to maintain.

ref: https://github.com/conda-forge/boost-cpp-feedstock/blob/bae4ccf3b9ea2be3b452de3628902b22919241f5/recipe/meta.yaml#L50-L91

File: recipe/meta.yaml [hunk 7]
```diff
@@ -8,14 +8,18 @@ source:
...
 build:
-    number: 0
-    skip: true    # [win]
+    number: 1
 
 requirements:
     build:
         - gcc     # [unix]
+        - {{ compiler('cxx') }}  # [win]
         - hdf5 1.8.18|1.8.18.*
```
Comment (jakirkham) [inline recipe/meta.yaml]: Please change to `hdf5 1.10.1|1.10.1.*`. Also please make this change below.

**Comment by jakirkham** [comment]
Couple minor things. Otherwise LGTM.

@dougalsutherland, thoughts?

**Comment by chambbj** [comment]
This should be closer to what you suggested @jakirkham

**Comment by djsutherland** [comment]
LGTM. Hitting merge.

Thanks a ton @chambbj! Will send PRs for `pyflann` and `cyflann` once this builds. :)

**Comment by jakirkham** [comment]
Great, thanks @chambbj. Also thanks for helping review @dougalsutherland.

**Pull request merged**